feat: add hint to run `rustup self` when err desc is `self`

Xerxes-2 opened the pull request:
Closes https://github.com/rust-lang/rustup/issues/3812.

djc commented on src/dist/mod.rs:
```diff
@@ -319,7 +320,11 @@ impl FromStr for ParsedToolchainDesc {
+            if desc_string.as_str() == "self" {
+                info!("If you meant to modify rustup itself, try run `rustup self`");
+            }
+            Err(RustupError::InvalidToolchainName(desc_string).into())
```
I think it would be better to do this near the top of the method? We can just check for `desc == "self"` and return the error immediately, which would save a bunch of work trying to parse.

rami3l replied:
@djc My suggestion is that since `FromStr` is used in many places other than `update` and `uninstall` (such as `install`), this will cause many false positives, and instead we should modify the `value_parser` for these two commands only.

djc replied:
But `self` is still invalid in all those places, right -- worst case this will provide a clearer error message?

rami3l replied:
@djc I'm afraid that will make the error message less clear. `rustup install self` doesn't make sense and if the info leads the user to `rustup self` they won't find anything related to `install` there.

djc replied:
Sure, but users are presumably also much less likely to invoke `rustup install self`. I still think it would be better to embed this check inside the `FromStr` impl where it is implicitly used in any future code that wants to consume toolchain names instead of remembering that we have to add an extra check everywhere, we can make the error message slightly more generic to make sure it doesn't lose clarity.

rami3l replied:
@djc I still feel uneasy about adding side effects in conversions like `FromStr`... @ChrisDenton What do you think?

djc replied:
I would argue that logging is pretty much "unobservable" as side effects go, so while I agree that in general side effects in something like a `FromStr` implementations could be a bit iffy, I think logging is an exception to this.

Would you have problems with adding logging to other kinds of "pure" functions?

rami3l replied:
@djc Actually... it's not about purity, it's about the fact that conversions can be used in various places across the codebase, many of which are not easily detectable via call hierarchy graph `rust-analyzer` gives us, and I'm afraid of this potentially adding a log line somewhere unexpected. Since the very nature of this feature is to add a hint, I guess it'd be better to get the hint only when we are explicitly expecting for it.

For example, `rustup uninstall` uses `ResolvableToolchainName`, and its conversions relies on the conversions defined for `PartialToolchainDesc` first, and that of `CustomToolchainName` as a fallback. Let's say we're still using https://github.com/rust-lang/rustup/commit/a743b3adc19ec949761a236fe484b0e34f0f44fb, and it so happens that there's a custom toolchain named `self`, then the first conversion will definitely fail. However, before the fallback happens, the log line is already printed out.

Xerxes-2 committed:
apply smart guess to `rustup update/uninstall self`

Use original error instead of rewriting

avoid duplicate `.to_string()`

remove unnecessary `.as_str()`

rami3l commented on src/cli/rustup_mode.rs:
```diff
@@ -258,6 +261,15 @@ enum RustupSubcmd {
+        if s == "self" {
+            info!("if you meant to update rustup itself, use `rustup self update`");
+        }
+        clap::Error::raw(ErrorKind::ValueValidation, e)
```
You should throw the original error instead of creating a new one. See https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect_err.

rami3l commented on src/toolchain.rs:
```diff
@@ -457,7 +457,13 @@ impl<'a> Toolchain<'a> {
                 } else {
+                    let name = name.to_string();
                     info!("no toolchain installed for '{name}'");
+                    if name.as_str() == "self" {
```
Will this work?

```suggestion
                    if &name == "self" {
```

Xerxes-2 replied:
Seems work even without `&`

rami3l replied:
> Seems work even without `&`

@Xerxes-2 That's better then.

rami3l reviewed (commented):
Thanks! IMHO there are two things left to do:

1. Squash all the feature changes together, as `rustup` encourages atomic commits, and the current changes are seemingly small enough to be squash into a single commit.
2. In a new commit, add some regression tests in `tests/suite/cli_misc.rs` to prevent future regressions.

Xerxes-2 commented:
> Thanks! IMHO there are two things left to do:
> 
> 1. Squash all the feature changes together, as `rustup` encourages atomic commits, and the current changes are seemingly small enough to be squash into a single commit.
> 2. In a new commit, add some regression tests in `tests/suite/cli_misc.rs` to prevent future regressions.

Fine, I might do it later cuz I'm now in a holiday.

Xerxes-2 commented:
@rami3l Can I close this PR and open a new one?

rami3l commented:
> @rami3l Can I close this PR and open a new one?

@Xerxes-2 Yes, but why you'd want to do that instead of doing an interactive rebase directly on this one?

Xerxes-2 commented:
> > @rami3l Can I close this PR and open a new one?
> 
> @Xerxes-2 Yes, but why you'd want to do that instead of doing an interactive rebase directly on this one?

Because I accidentally merged master into this branch, may not be able to squash together all the feature commits

rami3l commented:
> > > @rami3l Can I close this PR and open a new one?
> > 
> > @Xerxes-2 Yes, but why you'd want to do that instead of doing an interactive rebase directly on this one?
> 
> Because I accidentally merged master into this branch, may not be able to squash together all the feature commits

@Xerxes-2 AFAIK merge commits will be automatically dropped when doing an interactive rebase. See https://stackoverflow.com/a/4784082.

Xerxes-2 committed:
add regression tests for smart guess

rami3l commented:
@Xerxes-2 Thanks again for helping out!
